// tabletop-dice/index.js
class Dice {
  roll(maxValue, quantity = 1) {
    let values = []
    for (let a = 0; a < quantity; a++) {
      let randomValue = Math.round(Math.random() * (maxValue - 1)) + 1
      values.push(randomValue)
    }

    return quantity > 1 ? values : values[0]
  }

  d4(quantity = 1) {
    return this.roll(4, quantity)
  }

  d6(quantity = 1) {
    return this.roll(6, quantity)
  }

  d8(quantity = 1) {
    return this.roll(8, quantity)
  }

  d10(quantity = 1) {
    return this.roll(10, quantity)
  }

  d12(quantity = 1) {
    return this.roll(12, quantity)
  }

  d20(quantity = 1) {
    return this.roll(20, quantity)
  }

  advantage() {
    let values = this.d20(2);
    let highest = values.sort(function(a, b) {
      return b - a
    })[0]

    return {
      value: highest,
      rolls: values
    }
  }

  disadvantage() {
    let values = this.d20(2);
    let lowest = values.sort()[0]

    return {
      value: lowest,
      rolls: values
    }
  }

  percentage(quantity = 1) {
    return this.roll(100, quantity)
  }

  //Acceptable Inputs
  //!d4 !1d4 both result in rolling 1 4 sided die
  //!2d4 results in an array containing the results of 2 d4
  //!2d999 would roll 2 dice with 999 sides
  diceCommand(input) {
    var dVal; //number of sides on the die
    var dCount; //number of dice to roll
    var diceArr; //array of dVals for multiple rolls
    var total = 0; //total of diceArr components added together
    var output; //Output Message

    //Find Dice Value And Count
    dVal = input.substring(input.lastIndexOf("d") + 1); //Find last "d" in input, keeps value after
    dCount = input.substring(0, input.indexOf("d")); //Find letter "d" in input, keeps value before

    //If dCount isn't specified (input !d4, !d99, etc.) set to 1
    if (dCount == "" || dCount == undefined || dCount == null) {
      dCount = 1;
    }

    console.log("dVal | " + dVal);
    console.log("dCount | " + dCount);

    diceArr = Dice.roll(dVal, dCount);
    console.log("diceArr | " + diceArr);

    for (var i = 0; i < diceArr.length; i++) {
      total += diceArr[i];
    }

    if (dCount == 1) {
      output = " Rolled 1d" + dVal + " and got a " + diceArr;
    }
    else {
      output = " Rolled " + dCount + "d" + dVal + " and got " + diceArr + " for a total of " + total;
    }

    return output;
  }
}

module.exports = new Dice();
